subprojects/model-core/src/main/java/org/gradle/api/internal/provider/ValueSupplier.java

```diff
@@ -286,6 +333,68 @@ public String toString() {
         }
     }
 
+    class EmptySideEffect<T> implements SideEffect<T> {
+
+        public static boolean isEmpty(@Nullable SideEffect<?> sideEffect) {
+            return sideEffect == null || sideEffect == INSTANCE;
+        }
```

Comment:
Not sure if this method should be moved to the `SideEffect` interface. Apart from `ValueSupplier` it is used in `WithSideEffectProvider` as well, and may later be used in other places. Moving it to the `SideEffect` may allow to keep the `EmptySideEffect` class sort of private to the `ValueSupplier`.

Reply:
Yeah, this method seems to be better on the SideEffect interface.

Reply:
No longer relevant, because side effects are nullable now